nservicebus/nhibernate/custom-mappings.md

```diff
@@ -0,0 +1,126 @@
+---
+title: NHibernate Custom mapping with hbm.xml, attributes or fluent api
+summary: Create a custom mapping to change how NHibernate creates the database schema using different techniques.
+tags:
+ - NHibernate
+ - Mapping
+related:
+ - nservicebus/nhibernate/accessing-data
+---
+
+
+Sometimes the database schema that gets generated by `NServiceBus.NHibernate` for your saga data types is insufficient. For example, you might need to store strings with a different string length, want so use complex types, need to store DateTime values with a high precision or maybe want to tweak the eager versus lazy loading to improve performance. You will need to create a custom mapping to achieve this so that the generated default mapping will not be used.
+
+Custom mapping options:
+
+* Create a `*.hbm.xml` file for your saga data class
+* Use Fluent NHibernate (popular separate fluent api)
+* Use NHibernate.Mapping.Attributes
+* Loquacious Configuration (native fluent api)
+
+There are probably even other options but these are the most frequently used.
+
+NOTE: **Mention here if it is required to add custom mappings for all sagas**
+
+
+See [NHibernate configuration](configuration.md) on how you can pass this configuration object to the NServiceBus NHibernate configuration.
+
+NOTE: When you create a custom mapping then you are responsible for mapping the primary key and unique indexes correctly.
+
+
+NOTE: When using custom mappings you are responsible for creating unique indexes for all mapped saga properties. This is especially important when using both optimistic concurrency and a transaction isolation level different from serializable and allow concurrent processing of messages. Not adding this can result in multiple saga entities be persisted as the second insert will not fail when inserting a duplicate saga key (different from the primary key column) when multiple message  are processed concurrently that target the same saga instance.
+
+
+## Custom .hbm.xml mapping
+
+Using NHibernate mapping files is the original way to customize your mapping. It needs to be created and then either embedded as a resource in your assembly or available on the file system.
+
+NOTE: Using xml mappings is the default approach for NHibernate.
+
+Look at these pages to see how you can create these mappings:
```

Comment:
What about?

`For more information refer to the following documents on the official NHibernate documentation`